Added a cloud renderer that uploads geometry to the GPU.

Zaggy1024 opened the pull request:
#3559 updated to 1.11.2.

mezz commented:
Looks good, works as expected.

```
[18:22:07] [Client thread/INFO] [FML]: Vanilla clouds took 1.4857225ms per frame average
[18:22:17] [Client thread/INFO] [FML]: Forge clouds took 0.06850097ms per frame average
[18:22:27] [Client thread/INFO] [FML]: Vanilla clouds took 0.942308ms per frame average
[18:22:37] [Client thread/INFO] [FML]: Forge clouds took 0.05443495ms per frame average
[18:22:37] [Client thread/INFO] [FML]: Switched back to Forge clouds
```

mezz commented:
@RainWarrior please take a look

bs2609 commented:
This needs updating due to conflicts.

Zaggy1024 commented:
Done.

HenryLoenwind commented:
still conflicting...

Update: ok, ok: ~~still~~ conflicting *again* ;)

bs2609 commented:
bbc107de96a438496ae85223ca1b2a3d8478228f happened

Zaggy1024 commented:
Updated again.

mezz commented:
I think at this point there's no real question that the GPU cloud rendering is much faster, and if there is a question someone can open the F3 debug menu to see the time it takes.
Given that, I think you should remove the profiling patches so that you don't have to keep updating this PR when it conflicts. What do you think?

Zaggy1024 commented:
I agree that at this point the GPU rendering is much faster and that will be clearly shown by the in-game profiler, but as I understand it, the reason for the test mod was to make it easier to check if changes to the vanilla renderer make it faster than my code. If that's not necessary, though, I'll happily remove the profiler test.

mezz commented:
I think that the test mod was my request, but at this point with Forge's in-game configs working again, I think it would be simpler and less invasive to just check with F3. Just make sure that it's getting profiled properly so that comparison is possible.

Zaggy1024 commented:
It always has been profiled properly, in fact that's what allowed the test mod to work. :) I'll get on removing it when I get a chance.

Zaggy1024 committed:
Added a cloud renderer that uploads geometry to the GPU.

Zaggy1024 commented:
Done, there is now no profiling code. :)

Edit: To clarify, the vanilla cloud profiling section works for Forge's cloud renderer as well, so that can still be used to compare them.

mezz commented:
Cool, looks super clean now.

RainWarrior commented:
Why are .checkSettings() and .render() separate? Also, I would create and register the instance inside FMLClientHandler.baginMinecraftLoader, that way there's no need for the unchecked cast, and no need for separate init method and constructor. The whole patch inside RenderGlobal.renderClouds (old part too) can be replaced with 1 call to FMLClientHandler method - that way there's no need for the public instance of the cloud renderer either.

Zaggy1024 committed:
Moved cloud renderer hook and instance into FMLClientHandler.

Zaggy1024 commented:
.checkSettings() is separate so that it can run that code as little as possible (which without a settings change event is every client tick).

I've moved the init and hook into FMLClientHandler.

mezz commented:
`checkSettings`'s normal cast looks like it could be pretty quick, it's just checking booleans.
Can you clean up `checkSettings` a bit? The logic seems a bit odd. It seems like there are paths that could build or dispose multiple times.

mezz commented:
Any update @Zaggy1024 ?

Zaggy1024 committed:
Rewrote checkSettings to only dispose and build the cloud buffer when necessary.

Zaggy1024 commented:
I just rewrote checkSettings, should only mess with the buffer when necessary now. I'm not sure what cast you were talking about though.

Lemme know if it's ready and I can squash the commits down.

bs2609 commented on src/main/java/net/minecraftforge/client/CloudRenderer.java:
```diff
@@ -0,0 +1,466 @@
+    private final ResourceLocation texture = new ResourceLocation("textures/environment/clouds.png");
+
+    private int displayList = -1;
+    private net.minecraft.client.renderer.vertex.VertexBuffer vbo;
```
This can just be `VertexBuffer`

bs2609 commented on src/main/java/net/minecraftforge/client/CloudRenderer.java:
```diff
@@ -0,0 +1,466 @@
+        VertexBuffer buffer = tess.getBuffer();
+
+        if (OpenGlHelper.useVbo())
+            vbo = new net.minecraft.client.renderer.vertex.VertexBuffer(FORMAT);
```
as can this

bs2609 commented on src/main/java/net/minecraftforge/fml/client/FMLClientHandler.java:
```diff
@@ -1114,4 +1119,15 @@ public boolean isDisplayVSyncForced()
+
+    public boolean renderClouds(int cloudTicks, float partialTicks)
+    {
+        net.minecraftforge.client.IRenderHandler renderer = this.client.world.provider.getCloudRenderer();
```
Just `IRenderHandler` (use an import)

bs2609 reviewed (commented):
There's a couple of fully-qualified class names lurking in Forge code, probably from a former patch. For tidiness, just use simple names and add imports.

Zaggy1024 committed:
Changed fully qualified class name in FMLClientHandler to an import.

Zaggy1024 commented:
VertexBuffer has to be fully qualified because there are two classes with the same MCP name in use in the renderer.

Fixed the other one, that was copied over from a patch to a vanilla class.

bs2609 commented:
Ah right, I see. I missed the extra 'vertex' in the package name. Unfortunate but it can't be helped.

mezz commented:
Ah typo, I think I meant "case" and not "cast".

mezz commented:
Please create a 1.12 version of this PR, and leave this one open. Once the 1.12 version is accepted and merged, this one can be too.
@RainWarrior please take a look at this one.

Zaggy1024 committed:
Removed unused imports in ForgeModContainer.

Zaggy1024 committed:
Made cloud distance scale with render distance. (Not present in vanilla)
Fixed an error that occurs in loading if resources aren't reloaded after CloudRenderer is instantiated. (Normally only occurs in Forge 1.12)

Zaggy1024 commented:
If you think it's ready to go, I can squash it whenever you want.

1.12 version incoming in a minute.

bs2609 commented:
Is this up-to-date with the 1.12 version?

Zaggy1024 commented:
No, I made several changes since this PR. Why? Do you want to backport the changes?

williewillus commented:
Probably because it's in "Assigned" status (aka finished and pullable) but is out of sync with 1.12

mezz commented:
This was assigned to fry for review. "Assigned" to fry here just means "please ignore this, stale bot".
If you want this feature in 1.11 (or any lower version) just update this to match the 1.12 version and I will pull it.

m00nwtchr commented:
big yes to this :)
